src/CondenserDotNet.Configuration/Consul/ConsulConfigSource.cs

```diff
@@ -0,0 +1,123 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Options;
+using Newtonsoft.Json;
+
+namespace CondenserDotNet.Configuration.Consul
+{
+    public class ConsulConfigSource
+    {
+        private const string ConsulPath = "/";
+        private const char ConsulPathChar = '/';
+        private const char CorePath = ':';
+        private const string ConsulKeyPath = "/v1/kv/";
+        private const string IndexHeader = "X-Consul-Index";
+
+        private readonly string _agentAddress;
+        private readonly HttpClient _httpClient;
+        private readonly CancellationTokenSource _disposed = new CancellationTokenSource();
+
+        private IKeyParser _parser;
+
+        public ConsulConfigSource(IOptions<ConsulRegistryConfig> agentConfig)
+        {
+            var agentInfo = agentConfig?.Value ?? new ConsulRegistryConfig();
+            _agentAddress = $"http://{agentInfo.AgentAddress}:{agentInfo.AgentPort}";
+            _parser = agentInfo.KeyParser;
+            _httpClient = new HttpClient()
+            {
+                BaseAddress = new Uri(_agentAddress)
+            };
+        }
+
+        public string FormValidKey(string keyPath)
+        {
+            if (!keyPath.EndsWith(ConsulPath)) keyPath = keyPath + ConsulPath;
+
+            return keyPath;
+
+        }
+
+        public async Task<(bool success, Dictionary<string, string> dictionary)> GetKeysAsync(string keyPath)
+        {
+            var response = await _httpClient.GetAsync($"{ConsulKeyPath}{keyPath}?recurse");
+            if (!response.IsSuccessStatusCode)
+            {
+                return (false, null);
+            }
+
+            var dictionary = await BuildDictionaryAsync(keyPath, response);
+            return (true, dictionary);
+        }
+
+        public async Task WatchKeysAsync(string keyPath, Action<Dictionary<string, string>> onUpdate)
+        {
+            var consulIndex = "0";
+            var url = $"{ConsulKeyPath}{keyPath}?recurse&wait=300s&index=";
+            try { 
+                while (true)
+                {
+                    var response = await _httpClient.GetAsync(url + consulIndex, _disposed.Token);
+                    consulIndex = GetConsulIndex(response);
+                    if (!response.IsSuccessStatusCode)
```

Comment:
Same thing as we put in the service registry today, get the consul index into a temp var and check against the current one, if its not different repeat the loop with a continue.